Graphics/Implicit/ExtOpenScad/Eval/Constant.hs

```diff
@@ -52,52 +51,29 @@ import Graphics.Implicit.ExtOpenScad.Parser.Lexer (matchTok)
 addConstants :: [String] -> IO (VarLookup, [Message])
 addConstants constants = do
   path <- getCurrentDirectory
-  let scadOpts = ScadOpts False False
-  (_, CompState (varLookup, _, _, messages, _)) <- liftIO $ runStateT (execAssignments constants 0) (CompState (defaultObjects, [], path, [], scadOpts))
-  pure (varLookup, messages)
+  (_, s) <- liftIO . runStateT (execAssignments constants) $ CompState defaultObjects [] path [] opts
+  pure (varLookup s, messages s)
   where
-    execAssignments :: [String] -> Fastℕ  -> StateC ()
-    execAssignments [] _ = pure ()
-    execAssignments (assignment:xs) count = do
-      let
-        pos = SourcePosition count 1 "cmdline_constants"
-        show' err = showErrorMessages "or" "unknown parse error" "expecting" "unexpected" "end of input" (errorMessages err)
-      case parseAssignment "cmdline_constant" assignment of
-        Left e            -> addMessage SyntaxError pos $ show' e
-        Right (key, expr) -> do
-          res <- evalExpr pos expr
-          case matchPat key res of
-            Nothing -> pure ()
-            Just pat -> modifyVarLookup $ varUnion pat
-      execAssignments xs (count+1)
-    parseAssignment :: SourceName -> String -> Either ParseError (Pattern, Expr)
-    parseAssignment = parse assignment
-      where
-        assignment :: GenParser Char st (Pattern, Expr)
-        assignment = do
-          key <- patternMatcher
-          _ <- matchTok '='
-          expr <- expr0
-          pure (key, expr)
+    opts = ScadOpts False False
+    show' = showErrorMessages "or" "unknown parse error" "expecting" "unexpected" "end of input" . errorMessages
+    execAssignments :: [String] -> StateC Fastℕ
+    execAssignments = foldlM execAssignment 0
+    execAssignment count assignment = do
+        let pos = SourcePosition count 1 "cmdline_constants"
+            err = addMessage SyntaxError pos . show'
+            run (k, e) = evalExpr pos e >>= traverse_ (modifyVarLookup . varUnion) . matchPat k
+        either err run $ parseAssignment "cmdline_constant" assignment
+        pure $ count + 1
```

Comment:
The bit of computation we really want to nail down. This lets us focus on a single assignment at a time without having to worry about how it is connected to the rest.
traverse_ is doing the same computation here as the Maybe pattern match.